src/core/vtk/vtkFieldRef.ts

```diff
@@ -150,36 +171,59 @@ export function vtkFieldRef<T extends Maybe<vtkObject>>(
   });
 
   const onModified = batchForNextTask(() => {
-    if (unref(obj)?.isDeleted()) return;
-
-    // Special handling for array values that might have been mutated
-    if (lastValueIsArray) {
-      const currentValue = getter();
-      if (Array.isArray(currentValue)) {
-        const previousValue = lastValue;
-        // Check if array contents changed
-        if (
-          previousValue.length !== currentValue.length ||
-          previousValue.some(
-            (val: any, idx: number) => val !== currentValue[idx]
-          )
-        ) {
-          // Values changed, trigger the ref
-          triggerRef(ref);
-          return;
-        }
+    const currentVtkObj = unref(obj);
+    if (currentVtkObj?.isDeleted()) return;
+
+    const currentSourceValue = _sourceGet();
+    let valueChanged = false;
+
+    if (lastValueIsArray && Array.isArray(currentSourceValue)) {
+      if (
+        lastValue.length !== currentSourceValue.length ||
+        lastValue.some(
+          (val: any, idx: number) => val !== currentSourceValue[idx]
+        )
+      ) {
+        valueChanged = true;
       }
+    } else if (lastValue !== currentSourceValue) {
+      valueChanged = true;
     }
 
-    // Always trigger for non-array values
-    triggerRef(ref);
+    if (valueChanged) {
+      triggerRef(ref);
+    }
   });
 
-  ({ pause, resume } = onPausableVTKEvent(
-    obj as vtkObject,
-    'onModified',
-    onModified
-  ));
+  const setupVtkListener = (targetVtkObject: Maybe<vtkObject>) => {
+    currentPauseListener(); // Clean up previous listener's effect
+
+    if (targetVtkObject && !targetVtkObject.isDeleted()) {
+      const { pause: newPause, resume: newResume } = onPausableVTKEvent(
+        targetVtkObject,
+        'onModified',
+        onModified
+      );
+      currentPauseListener = newPause;
+      currentResumeListener = newResume;
+    } else {
+      currentPauseListener = () => {};
+      currentResumeListener = () => {};
+    }
+  };
+
+  if (isRef(obj)) {
+    watch(
+      obj,
+      (newVtkObj) => {
+        setupVtkListener(newVtkObj);
+        triggerRef(ref); // Trigger dependents as the source object has changed
+      },
+      { immediate: true }
+    );
+  } else {
+    setupVtkListener(unref(obj));
```

Comment:
This looks to be the equivalent of `onPausableVTKEvent` with a watcher to trigger the ref whenever `obj` changes.
- Not sure why the watcher is immediate. I don't think we need to trigger the ref right before returning it for the first time.
- The whole setupVtkListener just tears down (incorrectly, too; currentPauseListener() isn't stop()) and sets up onPausableVTKEvent whenever obj changes, when onPausableVTKEvent should be handling changes to obj internally.

```
pause, resume = onPausableVTKEvent(...)

watch(obj, () => triggerRef(ref))
```